### DISCUSSION TITLE
Column encryption with pgcrypto

### DISCUSSION THREAD

**Discussion by kmiyashiro** [supabase/supabase#627] [Questions] [answered]
I was looking into encrypting a sensitive column and read about the `pgcrypto` extension. I see `pgcrypto` is installed by default in supabase, but I was wondering how I would use it on my own tables or if it was something intended only for the auth table for now.

**Comment by ftonato** [comment, marked as answer]
Hey @kmiyashiro,

As you already know, **Supabase** allows the use of `pgcrypto`.

Although this is normally used for sensitive data (passwords), you can use it for whatever you want.

Below is an example of how to use it:


```sql
# Create table users
CREATE TABLE users (id SERIAL PRIMARY KEY,
                    email TEXT NOT NULL UNIQUE,
                    password TEXT NOT NULL);

# Insert a new user
INSERT INTO users (email, password)
VALUES ('kmiyashiro@mail.com', crypt('kmiyashiropassword', gen_salt('bf')));

# Fetch the user ID using his/her email and encrypted password
SELECT id
FROM users
WHERE email = 'kmiyashiro@mail.com'
  AND password = crypt('kmiyashiropassword', password);

# This will return one row with the correct user ID
# | id      |
# |---------|
# | 1       |
# | (1 row) |
```

**Note:** You can use this encryption for any field that you consider important.

I hope I've helped  🎸

**Reply by kmiyashiro** [reply]
Thanks for the reply! Yes that makes it clear how to use pgcrypto with a query, but how would I allow this to happen using supabase-js? Is there a way to read/write using pgcrypto automatically? Something similar to this: https://medium.com/@jianshi_94445/encrypt-decrypt-your-data-using-native-postgres-functions-with-sequelize-js-c04948d96833

**Reply by ftonato** [reply]
I can see your point, but unfortunately.
I don't think we have that yet (Could you confirm? @kiwicopple).

Anyway, it is an interesting suggestion that may make sense for the future.

Thanks for sharing  🚀

**Reply by kiwicopple** [reply]
You're right @ftonato - we don't have support for functions within the library yet, but this is something that Steve is thinking about for PostgREST. He is considering it for working with JSON data, but it will probably allow for overlap with other database functions.

For now, the only way you could achieve this is to use Stored Procedures (https://supabase.io/docs/client/rpc)

**Reply by kmiyashiro** [reply]
@kiwicopple got it, that might be a good workaround. I'll check it out!

**Reply by ClaudiuBadau** [reply]
This is great, however when I try to encrypt things in a trigger, it doesn't work. Am I doing something wrong?

This **works**
```
create or replace function public.handle_new_user() 
returns trigger as $$
begin
  insert into public.users (id, email)
  values (new.id, new.email);
  return new;
end;
$$ language plpgsql security definer;
```
This **doesn't work** - I get this error `{ message: 'Database error saving new user', status: 500 }`

```
create or replace function public.handle_new_user() 
returns trigger as $$
begin
  insert into public.users (id, email)
  values (new.id, pgp_sym_encrypt(new.email, 'KEY'));
  return new;
end;
$$ language plpgsql security definer;
```

The only difference is `pgp_sym_encrypt(new.email, 'KEY')`

I feel like I'm missing something. Any help would be appreciated 🙏

**Reply by rikur** [reply]
Subscribing, column encryption could be a nice middle ground between encryption at rest and e2ee.

**Reply by elhe26** [reply]
@kiwicopple any updates? This is a must-have feature. This could function as an alternative to multi-tenancy support.

**Reply by rikur** [reply]
@kiwicopple ping

**Reply by steve-chavez** [reply]
@rikur Can you explain what exactly is missing? What interface would you expect/want for encryption?

**Reply by rikur** [reply]
@steve-chavez I will test again to see if things have changed, but I want to use `pgcrypto` to encrypt a column and decrypt it on-demand via SELECT statements.

**Reply by Gerald-Gmainer** [reply]
@rikur @kmiyashiro kmiyashiro Are there any updates to use pgcrypto to encrypt and decrypt with supabase-js?

This is how I added data on the console
`update profiles set encrypted_field = pgp_sym_encrypt('hello', 'awesome_crypto_password');`

Selecting it on console works also
```
select pgp_sym_decrypt(encrypted_field, 'awesome_crypto_password') from profiles;
 pgp_sym_decrypt 
-----------------
 hello
(1 row)
```


I am trying to use the supabase-js client like this:
`let query = client.from('profiles').select(`pgp_sym_decrypt(encrypted_field, 'awesome_crypto_password'`)`

But I am getting this error
`column profiles.pgp_sym_decrypt does not exist`


Is there a way to fix it; besides RPCs?

**Reply by steve-chavez** [reply]
@Gerald-Gmainer You can use a computed column, it's a bit simpler than a full set returning function:

```sql
create function encrypted_field(profiles) returns bytea as $$
  select pgp_sym_decrypt($1.field, 'awesome_crypto_password');
$$ language sql;
```

```js
let query = client.from('profiles').select('encrypted_field, other_field, another')
```

**Reply by Gerald-Gmainer** [reply]
@steve-chavez Thank you for suggestion about the computed column.
Sorry, how I can use your function `encrypted_field` for the computed column?

Creating this function returns an error:
`ERROR:  type p does not exist`

**Reply by steve-chavez** [reply]
I've corrected the syntax ^

**Reply by Gerald-Gmainer** [reply]
@steve-chavez Thanks a lot!

**Reply by jsonpreet** [reply]
@steve-chavez I am getting error on function

```
CREATE TABLE public.encrypted_tasks (
  id serial primary key,
  title text null,
  user_id uuid null
)
```

```
create function encrypted_field(encrypted_tasks) returns bytea as $$
  select pgp_sym_decrypt($1.title, 'testkey');
$$ language sql;
```

```
ERROR:  42883: function pgp_sym_decrypt(text, unknown) does not exist

LINE 2:   select pgp_sym_decrypt($1.title, 'testkey');

HINT:  No function matches the given name and argument types. You might need to add explicit type casts.
```

**Reply by gcasesor** [reply]
i tried using pgp_sym_decrypt but got the same error --> ERROR:  42883: function pgp_sym_decrypt(text, unknown) does not exist.
Kindly advise.

**Reply by jakub-sekula** [reply]
I'm getting the same error here

**Reply by ivanbogdan-git** [reply]
Same error when using:
hashed_auth_code := encode(digest(id::TEXT || hash_salt, 'sha256'), 'hex');
function digest(text, unknown) does not exist.

Any update on this?

**Reply by ivanbogdan-git** [reply]
In my case, the solution was to remove the following line from the function:
```sql
SET search_path = public;
```
After digging deeper, I found out the reason this didn't work as expected was because pgcrypto was installed in the extensions schema (or namespace), not the public schema.

To check where pgcrypto was installed, I used the following script:

```sql
SELECT
    e.extname AS extension_name,
    n.nspname AS schema_name
FROM
    pg_extension e
JOIN
    pg_namespace n ON n.oid = e.extnamespace
WHERE
    e.extname = 'pgcrypto';
```
After confirming that pgcrypto was installed in the extensions schema, I updated the search path to include it:

```sql
SET search_path = public, extensions;
```
This resolved the issue by ensuring the function could access both the public schema and the extensions schema where pgcrypto was installed.

P.S. Someone somewhere mentioned that usign `public.digest` worked. But it didn't in my case.

**Reply by BeppeMarnell** [reply]
> In my case, the solution was to remove the following line from the function:
> 
> ```sql
> SET search_path = public;
> ```
> 
> After digging deeper, I found out the reason this didn't work as expected was because pgcrypto was installed in the extensions schema (or namespace), not the public schema.
> 
> To check where pgcrypto was installed, I used the following script:
> 
> ```sql
> SELECT
>     e.extname AS extension_name,
>     n.nspname AS schema_name
> FROM
>     pg_extension e
> JOIN
>     pg_namespace n ON n.oid = e.extnamespace
> WHERE
>     e.extname = 'pgcrypto';
> ```
> 
> After confirming that pgcrypto was installed in the extensions schema, I updated the search path to include it:
> 
> ```sql
> SET search_path = public, extensions;
> ```
> 
> This resolved the issue by ensuring the function could access both the public schema and the extensions schema where pgcrypto was installed.
> 
> P.S. Someone somewhere mentioned that usign `public.digest` worked. But it didn't in my case.



Following up on this issue and this response:
make sure you have installed pgcrypt in the extension and then you can use the `pgp_sym_encrypt/decrypt` functions like this, just adding the extension namespace:
```
extensions.pgp_sym_encrypt(bla,bla,bla)
extensions.pgp_sym_decrypt(bla,bla,bla)
```
